### ISSUE TITLE
Unnecessary cast for non-literal expressions

### ISSUE THREAD

**Issue by kornelski** [rust-lang/rust-clippy#8237]
### What it does

There's an `unnecessary_cast` lint for literals. Unnecessary casting can also happen for expressions. For example, I've refactored a codebase to remove uses of `c_int`, which left me with lots of `as _` conversions that were no longer doing anything.

I'm proposing this as a new lint, not an extension to the `unnecessary_cast`, because users may want to enable/disable this lint separately.

### Lint Name

unnecessary_cast_expression

### Category

complexity

### Advantage

Clarifies intent of the code.  Removes doubt whether the cast is intentional, or written because of confusion about expression's type.

### Drawbacks

`as _` is easier to read and write than `try_into().unwrap()`, and code that works with non-generic types that are changed by conditional compilation may prefer the terser syntax.

```rust
#[cfg(foo)]
type Int = u32;

#[cfg(not(foo))]
type Int = usize;

fn main() { takes_usize(my_int as _); }
```

### Example

```rust
fn takes_usize(_x: usize) {}

fn main() {
    takes_usize([1,2,3].len() as _);
}
```

**Comment by leonardo-m** [comment]
This looks good for rustc itself :-)

**Comment by TomFryersMidsummer** [comment]
I think this (now) exists in rustc, as [`trivial_casts`](https://doc.rust-lang.org/rustc/lints/listing/allowed-by-default.html#trivial-casts) and [`trivial_numeric_casts`](https://doc.rust-lang.org/rustc/lints/listing/allowed-by-default.html#trivial-numeric-casts).
